### PR TITLE
Fix "undo trap" by introducing "transient" attributes

### LINKED ISSUE #8119: "Undo trap": Avoid getting stuck in an editing state
**Undo trap:** The situation in which a user is "stuck" when continuously pressing Undo. The typical cause: as Undo is pressed and the content is reverted, some event somewhere causes a new undo level to be created. The result is that the user is unable to undo past a certain point, thereby losing access to previous editing states.

This issue aggregates all bugs stemming from that common problem:

- [x] Gallery: Conversion from raw (#6325, https://github.com/WordPress/gutenberg/pull/42001)
- [x] Embeds: Conversion from raw (#7323)
- [x] Image: Adding via upload (#3553)
- [x] File: Adding via drag-and-drop (#8155)

## Problem

At the root of these bugs is the fact that blocks need to manage a transition using specific ephemeral state:
- when converting from `[gallery ids="x,y,z"]`, a Gallery block needs to temporarily keep the IDs `x, y, z` of the images while it waits for API data necessary to the block (the URL and `alt` attribute for each image); once it receives the API data, it adds it to its attributes [resolved state];
- when adding an Image via file upload, the Image block initially keeps the image blob in order to render a pulsing image while waiting for upload conclusion; then, it updates its attributes to store the newly created URL for the uploaded image [resolved state];

Since the temporary states are encoded in the attributes themselves of the blocks, they automatically constitute an undo level of the editor history. Thus, the action of undoing will bring the blocks back to that temporary state, which itself triggers a transition "back" into the resolved state.

## Solutions

There are, as I see it, two approaches:

- Altogether avoid encoding transitional states in a block's attributes. Since this is a problem that is likely to affect third-party blocks as well, ideally the problem wouldn't be fixed _ad hoc_ but rather using a standardized method to express such states. It could be expressed in terms of creating a new block by passing it a "seed" for a future state (e.g. the seed of an image ID to later yield an image URL).
- Go in a very different direction and consider that certain "events" (term used loosely) in the application are made up of a sequence of action dispatches over time. So, for the constitution of a Gallery block converted from shortcode, the actions would be <block created with image IDs> and <block attributes updates with actual URLs>; these would fold into a single event, thus a single undo level. This approach is much more ambitious and likely to be overkill and less robust. I explain it at length in this comment: https://github.com/WordPress/gutenberg/pull/6325/#issuecomment-402242054

The first option seems to be worth exploring first. A quick stab at the second one — by hacking `withHistory` — doesn't inspire a lot of confidence.

### PR DESCRIPTION
**Pull request by mcsf**
Fixes #8119 

**Edit:** I'd like to explore another solution, but leaving this open for a while to discuss.

### Gallery: conversion from shortcode

gut-gallery

### Image: upload new image

gut-image

### Embed: convert from classic

gut-embed

### File: upload bad file

gut-file

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation.

### CODE REVIEW

**Review by aduth — CHANGES_REQUESTED** [review]
I need more context to understand the purpose its serving, but in the meantime I'll preemptively state my distaste for the exceptional treatment to "secret" attributes.

**Review by tofumatt — COMMENTED** [review]
Just piling on here: without suggesting a better solution, I worry about these too. While quite explicitly not part of the public API, I can imagine third-party folks finding and using these… and I also don't really like us using these kind of "secret" attributes in our own API.

I haven't seen it elsewhere in the codebase; it seems a bit un-React-y to me as well, but that's just, like, my opinion. 🤷‍♂️

**Comment by mcsf** [comment]
Just quickly replying from my phone: I don’t disagree with any of that, hopefully this is what the review is for. I’ve explored one or two other ways (see description of parent issue) to solve the problem, and none is entirely clean across the board. I still conceptually like the idea of “events comprised of multiple actions” better, but it required using Redux in an... original way. 

Anyway, I think a solution is very much needed for the issue. What alternative routes do you see? Silent attributes are a way to easily pass a piece of state to a block’s `edit` method, like a seed for the block to boot. I could imagine something requiring tweaking the `createBlock` interface, but I don’t know how we can move forward without interfering with the interfaces.

**Comment by mcsf** [comment]
I'd like to explore another solution, but leaving this open for a while to discuss.

**Comment by mcsf** [comment]
> I need more context to understand the purpose its serving, but in the meantime I'll preemptively state my distaste for the exceptional treatment to "secret" attributes.

@aduth, the parent issue should provide enough context — otherwise, ask me anything.

Does your distaste stem from the implicit nature of it (hidden attributes, no declaration in the type, a sketchy "convention" around the use of underscores), or is it more fundamentally about the idea of treating a certain kind of attribute differently?

More practically, if the solution would offer an interface like:

```js
attr1: {
  type: 'number',
  history: false, // doesn't persist
},
attr2: {
  type: 'string',
},
```

would that change anything? Because I don't care about this particular interface either, I wanted to see what the smallest change could look like. But the problem stands that some sort of state is needed for blocks that:
- can easily be passed when creating a new block (from conversion, etc.)
- is not committed to the undo history, i.e. so that it can never be accidentally replayed

Let me know, so I can either refactor this into something more palatable, or nix it in favour of something else.

**Comment by aduth** [comment]
I think the distaste comes from a perspective of looking to the attributes as the minimal set of values necessary to describe a block in its **saved state**.

From the parent issue, this reads as problematic in and of itself:

>Since the temporary states are encoded in the attributes themselves of the blocks

Temporary states should not be in the attributes. For what reason is the actual component state of the block's `edit` component insufficient?

I'd be okay with considering a change of attribute value as not being considered for history, but it's still a consideration of the specific implementation of how the `edit` component works at runtime, not part of the block type attributes definition. In other words, taking your example, I'd be more okay with `setAttributes( { attr1: 5 }, { history: false } );`

**Comment by mcsf** [comment]
> Temporary states should not be in the attributes. For what reason is the actual component state of the block's edit component insufficient?

The issue lies in tracking the circumstances of a block’s creation: when it’s the result of a transformation — in the broader sense — it can be incomplete. If we’re loading an existing post, all blocks therein are complete. If an existing block is being manipulated by the user, it can temporarily hold state in `edit`, then commit the proper data to its attributes. In contrast, if dragging an image file onto the editor canvas, pasting an embeddable URL, etc., the current Transformation API dictates that these events can be mapped to a block type, but the block type’s “runtime” (`edit`) ends up responsible for working out any missing values. 

Given this, the sane way to work with _attribute resolution_ should be to just wait inside a transformation’s callback until all the data requests have been resolved and only then instantiate a finalised block. The problem there, of course, is that we lose the ability to indicate progress (typically an incomplete block appears and will pulsate as data is loaded, etc.).

Thus, our options would be:

- (I'd run away from this option, but it's worth describing it.) Let the transformation know and do a lot more: parse its inputs; instantiate a block with missing attributes so that the block can display progress (e.g. an Image block instantiated with a `blob` file will be displaying a blinking image); take care of the data resolution (either up- or downloading state); once resolved, replace the previously instantiated block with a complete one. This not only implies a lot of changes in the block API, it also requires offloading onto it what has so far been handled directly in blocks' `edit` methods (e.g. loading images). Finally, it would easily break if a user manipulates the temporary block instance and we don't carry those changes into the final block instance properly.

- Retain the overall architecture, but support signalling to a new block that it should be aware of some missing data or be aware of some circumstances leading to its creation. That's where either silent attributes come in, *or* something like:

> I'd be more okay with setAttributes( { attr1: 5 }, { history: false } );

which is similar to my initial explorations (the more complex one using `withHistoryAmend` and the simpler, which was entirely equivalent).

**Comment by aduth** [comment]
>Given this, the sane way to work with attribute resolution should be to just wait inside a transformation’s callback until all the data requests have been resolved and only then instantiate a finalised block.

Yeah, one of my initial impressions might be that we add framework-level support for transformations to return a `Promise` which, until resolved, displays some progress state. The issues with this, as you allude to, are that (a) we'd need something generic enough to show and (b) it's inconsistent / redundant with a block's own implementation of its progress states.

_More thoughts to come..._

**Comment by aduth** [comment]
Relevant issues for problematic consequences of temporary states in attributes: #5936, #11565 

There could be some argument to make about supporting more formally this idea of states under which a block holds some temporary data, which in turn prevents a save from occurring.

**Comment by aduth** [comment]
Could this be something on an attribute definition, i.e. `transient: true`, which is a trigger both for (a) not creating a history entry and (b) not allowing save to occur until unset?

**Comment by aduth** [comment]
>but the interface would accommodate a passing of initialEditState to Image's edit method.

I like this proposal in that it preserves the idea of a two-way mapping between attributes and the persisted save form, but I worry it would require duplicating many of the mechanisms we've created around attributes (`setAttributes`, etc).

**Comment by mcsf** [comment]
> Could this be something on an attribute definition, i.e. `transient: true`, which is a trigger both for (a) not creating a history entry and (b) not allowing save to occur until unset?

Is this the same as the example in https://github.com/WordPress/gutenberg/pull/11504#issuecomment-437092082 ? If so, we can set out to build it.

> (b) not allowing save to occur until unset?

Hadn't thought about that, sounds interesting.

**Comment by aduth** [comment]
>Is this the same as the example in #11504 (comment) ? If so, we can set out to build it.

Effectively, yes, though less specific to the case of history and more to this idea of temporary states.

**Comment by mcsf** [comment]
> (b) not allowing save to occur until unset?

What about when some resolution fails and the user (rightfully) wishes to continue editing? How do we recover from that, save for deleting the offending(*) block? My default approach here would be to have the block serialiser systematically ignore transient attributes:

```diff
diff --git a/packages/blocks/src/api/serializer.js b/packages/blocks/src/api/serializer.js
index 443b68a14..c195ad4d4 100644
--- a/packages/blocks/src/api/serializer.js
+++ b/packages/blocks/src/api/serializer.js
@@ -152,6 +152,13 @@ export function getCommentAttributes( blockType, attributes ) {
 			return result;
 		}
 
+		// Ignore "transient" attributes.
+		//
+		// @see getTransientAttributes
+		if ( attributeSchema.transient === true ) {
+			return result;
+		}
+
 		// Ignore all attributes but the ones with an "undefined" source
 		// "undefined" source refers to attributes saved in the block comment.
 		if ( attributeSchema.source !== undefined ) {
```

I can picture situations where a user wants to edit the caption for an image that successfully resolved in that Gallery, even though some other image failed to load. The user shouldn't be prevented from saving that caption change nor should they lose that data.

Any reason not to go with that approach in `serializer.js`?

_(*): that phrase initially read "the blocking block"_

**Comment by mcsf** [comment]
In 748b4f9d8a9049d0449bfe9a3c522811c1538976 I switched to properly defined transient attributes.

> (b) not allowing save to occur until unset?

Thinking about this, I haven't yet updated all implicated blocks to unset their own transients once they're no longer needed, except for Gallery, to get a sense of what it looks like:

https://github.com/WordPress/gutenberg/pull/11504/commits/748b4f9d8a9049d0449bfe9a3c522811c1538976#diff-34f83402fe8954f15f8d404d01c9e4c9R128

File: packages/blocks/src/api/factory.js [hunk 1]
```diff
@@ -77,6 +84,38 @@ export function createBlock( name, attributes = {}, innerBlocks = [] ) {
...
+ * the media endpoint, and upon resolution update its own attributes to include
+ * image source and alt text. Thanks to its `transientImageIds` transient
+ * attribute, triggering Undo doesn't cause the block to retrigger this change.
+ *
+ * @param {Object} blockType  Block type.
+ * @param {Object} attributes Block attributes.
+ *
+ * @return {Object} "Transient" block attributes.
+ */
+export function getTransientAttributes( blockType, attributes ) {
```
Comment (aduth) [inline packages/blocks/src/api/factory.js]: Your branch is probably stale. Most functions accept `blockTypeOrName` as argument.

See: #11490

File: packages/blocks/src/api/serializer.js [hunk 2]
```diff
@@ -152,6 +152,13 @@ export function getCommentAttributes( blockType, attributes ) {
 			return result;
 		}
 
+		// Ignore "transient" attributes.
+		//
+		// @see getTransientAttributes
+		if ( attributeSchema.transient === true ) {
```
Comment (aduth) [inline packages/blocks/src/api/serializer.js]: Would truthy `if ( transient) {` suffice?

File: packages/editor/src/store/reducer.js [hunk 3]
```diff
@@ -235,6 +235,60 @@ export const editor = flow( [
 		resetTypes: [ 'SETUP_EDITOR_STATE' ],
 		ignoreTypes: [ 'RECEIVE_BLOCKS', 'RESET_POST', 'UPDATE_POST' ],
 		shouldOverwriteState,
+		serialize: ( present, past ) => {
```
Comment (aduth) [inline packages/editor/src/store/reducer.js]: What's this all about then?

**Comment by mcsf** [comment]
[emphasis mine in both quotes]

> Doesn't that then not do anything for us so far as resolving **#5936**, #11565 ?

Well, from the issue in question:

> An autosave should not take place if there are pending uploads, otherwise the **transient media information may be serialized** permanently to the post.

As this branch stands, serialisation of transients won't happen. What _can_ happen is that a block gets saved with partial information (e.g. missing an unresolved image), but at least it's not saved in a broken state of trying to fetch something.

> An autosave should not take place if there are pending uploads,

This makes some sense, but I worry about blocking users out of saving due to unforeseen failures — essentially, this gives a buggy block type the power to sabotage saving of an entire post. Don't you think so? (Coming from Calypso, where in the past I experienced breakage of the saving action due to an intermittent network connection, I'd do my best not to mess with saving too much.)

**Comment by aduth** [comment]
>I worry about blocking users out of saving due to unforeseen failures — essentially, this gives a buggy block type the power to sabotage saving of an entire post. Don't you think so?

I agree. I think also the alternative to save partial data is in its own way a broken experience. This seems one of those things which falls into the realm of the machine being unable to make a decision, and where we might even need to consider surfacing it to the user. For example, I could imagine a prompt about "The Image block has operations pending. Save anyways? [ Yes ] [ No ]", maybe highlighting the block in question.

File: packages/editor/src/store/reducer.js [hunk 4]
```diff
@@ -235,6 +235,60 @@ export const editor = flow( [
...
+			// Return the same reference unless change is necessary.
+			let nextState = present;
+
+			if ( past && past.blocks && present.blocks.byClientId === past.blocks.byClientId ) {
+				return nextState;
+			}
+
+			// For each changed block in the present, check for any transient
+			// attribute. If such an attribute is found, delete it from the
```
Comment (aduth) [inline packages/editor/src/store/reducer.js]: What's the issue with keeping it in state? Could it be enough to fall under the concept of `withHistory`'s "ignore" behavior?

Related prior art: caed8bf8f9c949c016a4ed14db85de7775c35f96 (#9403)

Comment (mcsf) [reply packages/editor/src/store/reducer.js]: As DM'd, this is problematic due to how we now expect to instantiate blocks with `createBlocks( { myNormalAttr, myTransientAttr } )`.

**Comment by aduth** [comment]
Summarizing direct message:

- The current proposal's `serialize` serves the purpose of treating transient as data which should never be restored in an undo, and should only exist in the present state of attributes.
- Much of my reluctance stems from a performance-related concern in the idea that we'd need to iterate all block's attributes for all blocks for all attribute updates dispatched to the store.
   - There is some shortcutting in place to relieve some of this worry.
- An option was considered in a combination of (a) partitioning incoming attributes updates to "fork" transients into their own, separate action update and (b) [implementing an ignore option](https://github.com/WordPress/gutenberg/pull/11504#discussion_r234037107) for the history enhancer.

**Comment by gziolo** [comment]
2 unit tests fail:

> [1] Test Suites: 1 failed, 1 skipped, 296 passed, 297 of 298 total
> [1] Tests:       2 failed, 1 skipped, 2633 passed, 2636 total
> [1] Snapshots:   1 failed, 160 passed, 161 total

They probably should be updated to reflect changes introduced.

**Comment by catehstn** [comment]
@aduth and @mcsf -- do you have an agreed on approach for this?
@mcsf -- do you think you are going to be able to return to this and update it ahead of 4.6?

**Comment by aduth** [comment]
To update, I'd iterated some on a separate branch toward the same end-goal here, exploring what it might look like to handle transient attribute updates from the block itself after the fact. The main hiccups involve issues around when and how to "flatten" history, mainly (a) the initial entry of the block having a temporary attribute value and (b) there being possible to be many states between the attribute value first being marked as transient and the transient state becoming completed (since an upload can take many actual seconds to complete).

The branch can be found here:

https://github.com/WordPress/gutenberg/compare/try/set-attributes-transient-option

It does reasonably well to avoid saving when an image upload is in-progress, but the aforementioned issues around history flattening become quite obvious in bugginess apparent when trying to Undo after an image upload had taken place.

It's also subject to the issue described at #12327, in that transient data is meant to live _outside_ the history of a block, but is only relevant so long as a reference to that block exists.

**Comment by catehstn** [comment]
Thanks @aduth, are you suggesting we close this PR in favour of yours? Or is your suggestion that we need to break this down into: resolve #12327, address undo bugginess, and then evaluate what's next?

**Comment by mcsf** [comment]
> are you suggesting we close this PR in favour of yours? Or is your suggestion that we need to break this down into: resolve #12327, address undo bugginess, and then evaluate what's next?

I just chatted with Andrew. We have no clear path forward yet, so I've opened and closed a PR from his branch, #12724, so that we can find it later, and I'll close this one as well. Meanwhile, the original issue, #8119, remains open.

**Pull request closed without merging**